Update README.md (#64)
replicatedhq/kots-field-labs: labs/lab08-airgap-lite/README.md

Lab 1.8: Air gap
=========================================

In this lab, we'll review how to perform installations and collect support bundles in Air Gap environments.

* **What you will do**:
    * Access and verify a single-node Air Gap setup via a bastion server
    * Learn how to install in an Air Gap environment using app manager
    * Create an SSH tunnel to configure an Air Gap instance
    * Perform an upgrade of an application in an Air Gap environment
    * Use the `kubectl support-bundle` CLI in an Air Gap environment
* **Who this is for**:

>     * Anyone who builds or maintains app manager applications (Full Stack/Devops/ Product Engineers)

Reviewer: ```suggestion
    * Anyone who builds or maintains app manager applications (Full Stack / Devops / Product Engineers)
```

From the Settings page, copy the Application Slug. You'll need this slug in a later step to set the `REPLICATED_APP`.

![kots-app-slug](img/application-slug.png)

***
## Instance overview

You will receive the IP of a jump box and the name of an Air Gap server.

For example:

```text
dx411-dex-lab08-airgap-lite-jump = 104.155.131.205
dx411-dex-lab08-airgap-lite
```

In general, the name of the private server will be the same as the jump box, with the characters `-jump` removed from the suffix.

For example, you can construct both instance names programmatically as:

```shell
${REPLICATED_APP}-lab08-airgap-lite-jump
```

```shell
${REPLICATED_APP}-lab08-airgap-lite
```

### Connecting to the Air Gap server

1. Set your application slug, the public IP of your jump box and your first name:
```shell
export JUMP_BOX_IP=...
export REPLICATED_APP=... # your app slug
export FIRST_NAME=... # your first name (lower case)
```

1. SSH into the Air Gap server using the following command:
  ```shell
  ssh -J ${FIRST_NAME}@${JUMP_BOX_IP} ${FIRST_NAME}@${REPLICATED_APP}-lab08-airgap-lite
  ```

> > **Note**: The `-J` option allows you to connect to the target host by first making a SSH connection to the jump host (`${JUMP_BOX_IP}`) described by the destination,  and then establishing a TCP forwarding to the ultimate destination (`${REPLICATED_APP}-lab08-airgap-lite`).

Reviewer: ```suggestion
    > **Note**: The `-J` option allows you to connect to the target host by first making a SSH connection to the jump host (`${JUMP_BOX_IP}`) described by the destination,  and then establishing a TCP forwarding to the ultimate destination (`${REPLICATED_APP}-lab08-airgap-lite`).
```

>   Optional: Instead of using the above command to SSH into the Air Gap server, you can perform multiple steps to achieve the same result:
>   ```shell
>   local> ssh ${FIRST_NAME}@${JUMP_BOX_IP}
>   jump> export REPLICATED_APP=...
>   jump> ssh ${REPLICATED_APP}-lab08-airgap-lite
>   ```

Reviewer: ```suggestion
    Optional: Instead of using the above command to SSH into the Air Gap server, you can perform multiple steps to achieve the same result:
    ```shell
    local> ssh ${FIRST_NAME}@${JUMP_BOX_IP}
    jump> export REPLICATED_APP=...
    jump> ssh ${REPLICATED_APP}-lab08-airgap-lite
    ```
```

Lab 1.8: Air gap
=========================================

In this lab, we'll review how to perform installations and collect support bundles in Air Gap environments.

* **What you will do**:
    * Access and verify a single-node Air Gap setup via a bastion server
    * Learn how to install in an Air Gap environment using app manager
    * Create an SSH tunnel to configure an Air Gap instance
    * Perform an upgrade of an application in an Air Gap environment
    * Use the `kubectl support-bundle` CLI in an Air Gap environment

> 

Reviewer: Is there a reason to add a newline? Markdown rendering isn't going to add a newline.

Author: I was doing it to stay consistent with the other items in the list. But it's fine if we want to keep it as is

* **Who this is for**:
    * Anyone who builds or maintains app manager applications (Full Stack/Devops/ Product Engineers)

> 

Reviewer: Same as line 12.

* **Prerequisites**:
    * Basic working knowledge of Kubernetes

> 

Reviewer: Same as line 12

* **Outcomes**:
    * You can deliver an app manager application into an Air Gap environment
    * You can build confidence in performing upgrades and troubleshooting in Air Gap environments

## Overview

To practice performing an Air Gap installation from scratch, you will start this lab with a bare Air Gap server with no app manager installation. Then we'll explore how some of the support techniques differ between online and Air Gap environments.

***
## Airgap workflow overview

First, we'll push a release and Replicated's Air Gap builder will prepare an Air Gap bundle.

![airgap-slide-1](img/airgap-slide-1.png)

> Next, we'll collect a license file, a download link, and a public Kubernetes installer bundle.

Reviewer: ```suggestion
Next, we'll collect a license file, a download link and a public Kubernetes installer bundle.
```

Author: Oxford comma for the win haha

![airgap-slide-3](img/airgap-slide-3.png)

The above diagram shows a three node cluster, but this lab will only use a single node cluster.
The app manager bundle will be moved onto the server via SCP, while
the app bundle and license file will be uploaded via a browser UI through an SSH tunnel.

***
## Getting started

You will receive an invite to log into https://vendor.replicated.com, accept this invite and set your password.

From the Settings page, copy the Application Slug. You'll need this slug in a later step to set the `REPLICATED_APP`.

![kots-app-slug](img/application-slug.png)

***
## Instance overview

You will receive the IP of a jump box and the name of an Air Gap server.

For example:

```text
dx411-dex-lab08-airgap-lite-jump = 104.155.131.205
dx411-dex-lab08-airgap-lite
```

In general, the name of the private server will be the same as the jump box, with the characters `-jump` removed from the suffix.

For example, you can construct both instance names programmatically as:

```shell
${REPLICATED_APP}-lab08-airgap-lite-jump
```

```shell
${REPLICATED_APP}-lab08-airgap-lite
```

### Connecting to the Air Gap server

> 1. Set your application slug, the public IP of your jump box and your first name:

Reviewer: If we use a numbered list, I suggest we indent also the code blocks below.

```shell
export JUMP_BOX_IP=...
export REPLICATED_APP=... # your app slug
export FIRST_NAME=... # your first name (lower case)
```

> 2. SSH into the Air Gap server using the following command:

Reviewer: ```suggestion
1. SSH into the Air Gap server using the following command:
```
Can we let markdown figure out the numbering. If you keep all of them at `1.`, markdown rendering will figure it out.

Reviewer: Same as above. I suggest we indent the below code block.

```shell
ssh -J ${FIRST_NAME}@${JUMP_BOX_IP} ${FIRST_NAME}@${REPLICATED_APP}-lab08-airgap-lite
```
> **Note**: The `-J` option allows you to connect to the target host by first making a SSH connection to the jump host (`${JUMP_BOX_IP}`) described by the destination,  and then establishing a TCP forwarding to the ultimate destination (`${REPLICATED_APP}-lab08-airgap-lite`).

  Optional: Instead of using the above command to SSH into the Air Gap server, you can perform multiple steps to achieve the same result:
  ```shell
  local> ssh ${FIRST_NAME}@${JUMP_BOX_IP}
  jump> export REPLICATED_APP=...
  jump> ssh ${REPLICATED_APP}-lab08-airgap-lite
  ```

> When you are on the Air Gap server, you can verify that the server does not have internet access. When you've confirmed no internet access exists, you can copy the below command and proceed to the next section.

Reviewer: ```suggestion
When you are on the Air Gap server, you can verify that the server does not have internet access. When you've confirmed no internet access exists, you can ctrl+C the below command and proceed to the next section.
```

***
## Moving Assets into place
To perform an Air Gap installation, you will need to collect the following assets:

* A license with the Air Gap entitlement enabled
* An Air Gap bundle containing the Kubernetes installer cluster components **
* An Air Gap bundle containing the application components **

> ** **Note**: These are separate artifacts to cut down on bundle size during upgrade scenarios where only the application version
is changing and no changes are needed to the underlying cluster.

> If you haven't already, you can log out of the Air Gap instance with `exit` or ctrl+D.

Reviewer: Maybe we should move this to the previous section? The next step is about getting back on the jumpstation, by using ssh from your workstation. And if they used the multistep aproach to get on the airgap instance, one `exit` will not bring them back onto the workstation.

#### Starting the Kubernetes installer bundle download

You will SSH into your jump box (the one with the public IP) `ssh kots@<jump box IP address>`, and download the Kubernetes installer bundle.

##### Enabling air gap for a customer

Enable Air Gap for the `lab8` customer and click **Save Changes**:

![enable-airgap](./img/airgap-customer-enable.png)

##### Downloading air gap assets
Scroll to the bottom of the page to the **Download Portal** section.

![download-portal](img/airgap-customer-portal.png)

1. Click **Generate new password** and save the generated password somewhere in your notes.

> 2. Click the **View download portal** link to open the download portal.

Reviewer: ```suggestion
1. Click the **View download portal** link to open the download portal.
```

>   > **Note**: This is a link you would usually send to your customer, so from this point forward in the lab we will be wearing our "end user" hat.

Reviewer: ```suggestion
    > **Note**: This is a link you would usually send to your customer, so from this point forward in the lab we will be wearing our "end user" hat.
```

> 3. Replace the URL below with the one you get from the download portal.

Reviewer: ```suggestion
1. Replace the URL below with the one you get from the download portal.
```

> **Note**: For a production application, Air Gap releases will be built automatically on the **Stable** channel.

For this lab, since we're working off the `lab08-airgap-lite` channel, you'll want to enable Air Gap builds on that channel.

You can check the build status by navigating to the **Release history** for the channel.

![release-history](img/channel-release-history.png)

You can build individual bundles on the **Release history** page, but you'll likely want to edit the channel and enable the option for all releases. Next to **Automatically create airgap builds for all releases in this channel**, slide the toggle swith to `on` to enable this option.

![edit-channel](img/channel-edit-info-btn.png)

![auto-build](img/channel-enable-airgap.png)

You should see all the bundles building or built on the **Release history** page:

![airgap-built](img/airgap-builds.png)

#### Continue downloading air gap assets
To continue downloading the air gap assets, navigate back to the download portal where we previously got the Kubernetes installer bundle URL.

1. Navigate to the **Embedded cluster** option and review the three downloadable assets.

>   ![download-portal-view](img/download-portal-view.png)

Reviewer: ```suggestion
    ![download-portal-view](img/download-portal-view.png)
```

> 2. Download the license file, but **do not** download the Kubernetes installer (kURL) bundle. This download was already started on the server in a previous step above.

Reviewer: ```suggestion
1. Download the license file, but **do not** download the Kubernetes installer (kURL) bundle. This download was already started on the server in a previous step above.
```

> 3. Download the `Latest Lab 1.8: Airgap Bundle` to your workstation.

Reviewer: ```suggestion
1. Download the `Latest Lab 1.8: Airgap Bundle` to your workstation.
```

> 4. SSH to your jump box (the one with the public IP) `ssh kots@<jump box IP address>` and check the download of the Kubernetes installer (kURL) bundle.

Reviewer: ```suggestion
1. SSH to your jump box (the one with the public IP) `ssh kots@<jump box IP address>` and check the download of the Kubernetes installer (kURL) bundle.
```

> 5. Replace the URL with the one you copied above.

Reviewer: ```suggestion
1. Replace the URL with the one you copied above.
```

```

It should be finished now, so you can copy it to the Air Gap server.
You can use the DNS name in this lab, as described in the [Instance Overview](#instance-overview).

```text
kots@dx411-dex-lab08-airgap-lite-jump ~$ scp kurlbundle.tar.gz ${REPLICATED_APP}-lab08-airgap-lite:~

```

> **Note**: -- we use SCP via an SSH tunnel in this lab, but the Air Gap methods in this lab also extend to more locked down environments where, for example, physical media is required to move assets into the datacenter.

The next step is to SSH all the way to Air Gap node. If you still have a shell on your jump box, you can use the instance name:

```text
kots@dx411-dex-lab08-airgap-lite-jump ~$ ssh ${REPLICATED_APP}-lab08-airgap-lite
```

Otherwise, you can use the one below:

```shell
ssh -J ${FIRST_NAME}@${JUMP_BOX_IP} ${FIRST_NAME}@${REPLICATED_APP}-lab08-airgap-lite
```

When you are on the Air Gap node, untar the bundle and run the install script with the `airgap` flag.
The Kubernetes installer (kURL) install flags are documented [in the kurl.sh docs](https://kurl.sh/docs/install-with-kurl/advanced-options).

```shell
tar xvf kurlbundle.tar.gz
sudo bash install.sh airgap
```

> At the end, you should see an `Installation Complete` message. Because this instance is an Air Gap, you must create a port forward to access the UI from your workstation in the next step.

Reviewer: ```suggestion
At the end, you should see an `Installation Complete` message. Because this instance is in Air Gap mode, you must create a port forward to access the UI from your workstation in the next step.
```
`an Air Gap` sounds weird? Or just me?

ssh -NL 8800:${REPLICATED_APP}-lab08-airgap-lite:8800 -L 8888:${REPLICATED_APP}-lab08-airgap-lite:8888 ${FIRST_NAME}@${JUMP_BOX_IP}
```

 At this point, Kubernetes and the Admin Console are running inside the air gapped server, but the application isn't deployed yet.

To complete the installation:
1. Visit http://localhost:8800 in your browser.

> 2. Click **Continue to Setup** in the browser to navigate to the secure Admin Console.

Reviewer: ```suggestion
1. Click **Continue to Setup** in the browser to navigate to the secure Admin Console.
```

>   ![kots-tls-wanring](img/kots-tls-warning.png)

Reviewer: ```suggestion
    ![kots-tls-wanring](img/kots-tls-warning.png)
```

> 3. Click **Skip & continue** to Accept the insecure certificate in the admin console.

Reviewer: ```suggestion
1. Click **Skip & continue** to Accept the insecure certificate in the admin console.
```

> > **Note**: For production installations we recommend uploading a trusted cert and key, but for this tutorial we will proceed with the self-signed cert.

Reviewer: ```suggestion
    > **Note**: For production installations we recommend uploading a trusted cert and key, but for this tutorial we will proceed with the self-signed cert.
```

>   ![Console TLS](img/admin-console-tls.png)

Reviewer: ```suggestion
    ![Console TLS](img/admin-console-tls.png)
```

> 4. On the login screen, paste the password noted previously on the `Installation Complete` screen. The password is shown in the output from the installation script.

Reviewer: ```suggestion
1. On the login screen, paste the password noted previously on the `Installation Complete` screen. The password is shown in the output from the installation script.
```

>   ![Log In](img/admin-console-login.png)

Reviewer: ```suggestion
    ![Log In](img/admin-console-login.png)
```

> Until this point, this server is just running Kubernetes and the `kotsadm` containers.

Reviewer: ```suggestion
    Until this point, this server is just running Kubernetes and the `kotsadm` containers.
```

> The next step is to upload a license file so app manager can validate which application is authorized to be deployed. You must use the license file we downloaded earlier.

Reviewer: ```suggestion
    The next step is to upload a license file so app manager can validate which application is authorized to be deployed. You must use the license file we downloaded earlier.
```

> 5. Click **Upload** and select your `.yaml` file to continue. You can also drag and drop the license file from a file browser. After you upload your license, you'll be greeted with an Airgap Upload screen.

Reviewer: ```suggestion
1. Click **Upload** and select your `.yaml` file to continue. You can also drag and drop the license file from a file browser. After you upload your license, you'll be greeted with an Airgap Upload screen.
```

>   ![Upload License](img/upload-license.png)

Reviewer: ```suggestion
    ![Upload License](img/upload-license.png)
```

> 6. Select **choose a bundle to upload** and use the application bundle that you

Reviewer: ```suggestion
1. Select **choose a bundle to upload** and use the application bundle that you
```

downloaded to your workstation using the customer portal. Click **Upload Air Gap bundle** to continue the upload process.

>   ![airgap-upload](img/airgap-upload.png)

Reviewer: ```suggestion
    ![airgap-upload](img/airgap-upload.png)
```

            - "1"
```

You must deploy a new release in order to fix this error.

***
## Deploying a new version

As part of the lab setup, a new release was created in the Vendor Portal with the fix. To make the release available:
1. Navigate to **Sequence 2** under **All releases**.  

> 2. Click **Promote**.

Reviewer: ```suggestion
1. Click **Promote**.
```

> 3. Select the `lab08-airgap-lite` channel as the channel to promote the release to.

Reviewer: ```suggestion
1. Select the `lab08-airgap-lite` channel as the channel to promote the release to.
```

>   ![app-down](img/promote-sequence-2.png)

Reviewer: ```suggestion
    ![app-down](img/promote-sequence-2.png)
```

               name: html
```

When the promotion is done, you should be able to navigate back to the customer download portal we accessed from the customer page.
Scrolling to the bottom, you can click **Show older bundles** to see the history of the releases on the `lab08-airgap` channel.

The new release may take a minute or two to build. You can refresh the page until you see a release with a timestamp that matches when you promoted your release.

![download-portal-more](img/download-portal-more.png)

When you've downloaded the new version, you can now deploy it:

> 1.  In the app manager Admin Console select **Version History**, click **Upload a new version**, and select your bundle.

Reviewer: ```suggestion
1. In the app manager Admin Console select **Version History**, click **Upload a new version**, and select your bundle.
```

>   ![airgap-new-upload](img/airgap-new-upload.png)

Reviewer: ```suggestion
    ![airgap-new-upload](img/airgap-new-upload.png)
```

>   You will see the bundle uploaded and you will have the option to deploy it once the preflight checks are complete.

Reviewer: ```suggestion
    You will see the bundle uploaded and you will have the option to deploy it once the preflight checks are complete.
```

> 2. Click **Deploy** to perform the upgrade.

Reviewer: ```suggestion
1. Click **Deploy** to perform the upgrade.
```

> 3. Click **Application** to navigate back to the main landing page. The app should now show the **Ready** status on the main dashboard.

Reviewer: ```suggestion
1. Click **Application** to navigate back to the main landing page. The app should now show the **Ready** status on the main dashboard.
```

> 4. To access the application select **Open Lab 8**.

Reviewer: ```suggestion
1. To access the application select **Open Lab 8**.
```

If you finish the lab early you can:

1. Experiment with copying the CLI-generated bundle off the server and uploading it to https://vendor.replicated.com.

> 2. Experiment with expanding or building your own `support-bundle.yaml` and using it to collect other information about the host.

Reviewer: ```suggestion
1. Experiment with expanding or building your own `support-bundle.yaml` and using it to collect other information about the host.
```
